src/ray/common/scheduling/cluster_resource_scheduler.cc

```diff
@@ -487,3 +621,225 @@ std::string ClusterResourceScheduler::DebugString(void) {
+  if (unit_instances) {
+    int64_t num_instances = static_cast<int64_t>(total);
+    instance_list->total.resize(num_instances);
+    instance_list->available.resize(num_instances);
+    for (size_t i = 0; i < num_instances; i++) {
+      instance_list->total[i] = instance_list->available[i] = 1.0;
+    };
+  } else {
+    instance_list->total.resize(1);
+    instance_list->available.resize(1);
+    instance_list->total[0] = instance_list->available[0] = total;
+  }
+}
+
+void ClusterResourceScheduler::InitLocalResources(const NodeResources &node_resources) {
+  local_resources_.predefined_resources.resize(PredefinedResources_MAX);
+
+  for (size_t i = 0; i < PredefinedResources_MAX; i++) {
+    if (node_resources.predefined_resources[i].total > 0) {
+      InitResourceInstances(
+          node_resources.predefined_resources[i].total,
+          (UnitInstanceResources.find(i) != UnitInstanceResources.end()),
+          &local_resources_.predefined_resources[i]);
+    }
+  }
+
+  if (node_resources.custom_resources.size() == 0) {
+    return;
+  }
+
+  for (auto it = node_resources.custom_resources.begin();
+       it != node_resources.custom_resources.end(); ++it) {
+    if (it->second.total > 0) {
+      ResourceInstanceCapacities instance_list;
+      InitResourceInstances(it->second.total, false, &instance_list);
+      local_resources_.custom_resources.emplace(it->first, instance_list);
+    }
+  }
+}
+
+void ClusterResourceScheduler::AddAvailableResourceInstances(
+    std::vector<double> available,
+    ResourceInstanceCapacities *resource_instances /* return */) {
+  for (size_t i = 0; i < available.size(); i++) {
+    resource_instances->available[i] = std::min(
+        resource_instances->available[i] + available[i], resource_instances->total[i]);
+  }
+}
+
+void ClusterResourceScheduler::SubtractAvailableResourceInstances(
+    std::vector<double> available,
+    ResourceInstanceCapacities *resource_instances /* return */) {
+  RAY_CHECK(available.size() == resource_instances->available.size());
+
+  for (size_t i = 0; i < available.size(); i++) {
+    resource_instances->available[i] =
+        std::max(resource_instances->available[i] - available[i], 0.);
+  }
+}
+
+bool ClusterResourceScheduler::AllocateResourceInstances(
+    double demand, bool soft, std::vector<double> &available,
+    std::vector<double> *allocation /* return */) {
+  allocation->resize(available.size());
+
+  if (available.size() == 1) {
+    // This resource has just an instance.
+    if (available[0] >= demand) {
+      available[0] -= demand;
+      (*allocation)[0] = demand;
+      return true;
+    } else {
+      if (soft) {
+        available[0] = 0;
+        return true;
+      }
+      // Not enough capacity.
+      return false;
+    }
+  }
+
+  // If resources has multiple instances, each instance has total capcity of 1.
+  // If demand is greater than 1., allocate full unit-capacity instances until the
+  // remaining demand is fractional.
```

Comment:
extend comment to explain why we break once it's fractional/what happens afterward?